### ISSUE TITLE
Facebook Messenger doesn't play nicely with AltTab

### ISSUE THREAD

**Issue by szhu** [lwouis/alt-tab-macos#3692]
**Describe the bug**

The official Facebook Messenger app has some issues with AltTab. AltTab is great with all other apps I use, so the root cause of these issues is likely with Messenger itself.

I wanted to report the bugs anyway:
1. To double-check that these bugs are indeed experienced by other people,
2. To see if anyone has come up with any workarounds, and/or
3. To see if there are any low-effort changes to AltTab that can be made to mitigate the issues.

There are two bugs described below. Let me know if it would be helpful to split them into two separate issues.

## Bug 1: AltTab sometimes doesn't foreground the main Messenger window

After the Messenger main window has been closed once, focusing it with AltTab doesn't bring it to the foreground.

**Screenshots / video**

https://github.com/user-attachments/assets/84f0f11b-4185-4f28-aa63-e4eab94616f2

**Steps to reproduce the bug**

1. Launch AltTab with default settings.
2. Launch Messenger.
   - Result: The Messenger main window appears.
3. Click the red 'X'.
   - Result: The Messenger main window disappears, and is no longer in the AltTab list.
4. Click Messenger in the Dock.
   - Result: The Messenger main window appears again.
5. Click another window.
   - Result: The other window becomes focused and in the foreground.
6. Press the AltTab shortcut and hold down the modifier key.
   - Result: The Messenger window is selected.
7. Let go of the modifier key.
   - Expected result: The Messenger main window becomes focused and in the foreground.
   - Actual result: The Messenger app becomes focused but not in the foreground.

## Bug 2: Sometimes blank "Media Viewer" windows appear in the AltTab list

After the Messenger Media Viewer window has been closed, it still appears in the AltTab list when the Messenger app is hidden. Each time a photo is opened, it creates another such phantom window.

**Screenshots / video**

https://github.com/user-attachments/assets/d506bdf0-6e10-4c41-8661-63a300d14a61

**Steps to reproduce the bug**

1. Launch AltTab with default settings.
2. Launch Messenger.
   - The Messenger main window appears.
3. Find a photo in any chat and click on it.
   - Result: The photo opens in a window titled "Media Viewer".
4. Click the red 'X'.
   - Result: The Media Viewer window disappears, and is no longer in the AltTab list.
5. Press Cmd-H.
   - Result: The Messenger main window disappears, and another app becomes focused.
6. Press the AltTab shortcut and hold the modifier key.
   - Expected result: There should be one Messenger window in the list: the main window.
   - Actual result: The main window is in the list, but so is a blank Media Viewer window. The number of Media Viewer windows there are corresponds to how many times a photo was opened in step 4.
7. Focus any Messenger window. Press the AltTab shortcut and hold the modifier key.
   - Result: The Media viewer windows are no longer in the AltTab list.

**Your environment**

* AltTab version: 7.0.2
* macOS version: 15.1 Beta (24B5035e)
* Other relevant info: N/A (no external devices connected)

**Comment by lwouis** [comment]
Hi,

Thank you for sharing these issues.

I couldn't reproduce the first issue. It works well in my tests. I wonder if the issue is not with Arc. Does the same thing happen for you if the window on top is TextEdit for instance?

I couldn't reproduce the second issue either. I don't get these phantom windows. Are you on the latest version Messenger? I tested with v220.1.0.5.212.

Thank you 🙇

**Comment by szhu** [comment]
Thanks for trying to reproduce it. The issue does not seem related to Arc; I tried with Finder and TextEdit instead of Arc (and with Arc not running) and ran into the same issue.

I tried with a new user account and it became impossible to reproduce both bugs on a fresh setup. I downloaded some of the apps I normally use, and was eventually able to reproduce the bug again, but not reliably.

This bug appears to be a trickier to narrow down that I had previously thought! I'll spend some more time narrowing a reliably reproducible test case and will report back.

**Comment by eskayn** [comment]
I am in quite a different situation but having a similar window not foreground issue on 15.0.1（24A348）with both AltTab 7.0.2 and 6.73.0. 

When I use a software called [XLD](https://tmkk.undo.jp/xld/index_e.html) to open a disc, it will first read the disc by showing a progress bar window. After it is finished the progress bar window disappears and there will be a new XLD window pops up onto the foreground(this is how XLD works). 
After that AltTab can not foreground the popped-up XLD window even choose it in AltTab. 
I tried a few times but I also found the bug seems to be tricky when I restart or use a new user the issue is gone but it returns after several trials.

**Comment by SarcevicAntonio** [comment]
I also have the not appearing in foreground issue on some apps, namely Slack and Telegram. A restart of the app fixes it for a bit, but after a while the issue reappears. Maybe thats why reproducing it is a bit tricky.

**Comment by champagnealexandre** [comment]
Can also confirm this with Facebook Messenger and Slack, among other apps which I can’t think of right now. Switching to these apps with Alfred works (i.e., brings them to the foreground) but not with Alt-Tab. I was wondering if this was an Electron issue but other Electron apps (e.g., the new Claude AI app) work fine. Maybe only some versions are affected.

**Comment by lwouis** [comment]
I'm not familiar with how Alfred focuses windows. I downloaded it to trial but couldn't figure it out. From [this page](https://alfred.app/workflows/alfredapp/window-switcher/), it suggest they only handle same-Space focus.

The implementation which AltTab uses is quite tricky. It allows focus in edge-cases such as a minimized window on another Space. This is not possible with public APIs, so AltTab uses tricks. As far as I know, it's the best method known currently.

If anyone knows a better implementation, I'd love to hear about it 👍

**Comment by champagnealexandre** [comment]
Yeah actually with Alfred I’m simply using its shortcut to bring it up (its "search box") then type the name of the app I wanna bring forward. It’s more or less equivalent to launching the app (from spotlight or the command line) so if it’s already open this will bring it to the foreground.

One thing that’s weird is that some apps e.g. Facebook Messenger worked until recently. Maybe the update to macOS Sequoia ruined it or something?

**Comment by lwouis** [comment]
Activating an app is quite different from focusing one window of that app.

AltTab focuses 1 window. For instance, if an app has 10 open window, AltTab will only bring 1 of them to the foreground.

if you use the native app switcher and switch to the app, all 10 of its windows will be brought forewords. I suggest you try it to see for yourself the difference.

If Already activates apps then they took the easy route. Focusing one window is really hard, as i explained above. We need to call private APIs

**Comment by filipef101** [comment]
Same thing with slack, and other apps as mentioned. I personally would be happy if we get an option to specify some apps that get focused on app instead of window. That would solve it :) usally slack, telegram, fb, its always 1 window anyway

**Comment by filipef101** [comment]
This might be a common thing with electron apps. It does seem it gets focused, but the window is not brought into foreground 🤔

**Comment by filipef101** [comment]
A possible hack could be, check if the app only has 1 window. if one window do the regular open app, or execute the bring all to front (speaking abstractly). Possibly do this for certain whitelisted apps if you want to support on the codebase specific app hacks :/ 
Should be trivial to replicate with slack. The main window never gets focused with alt tab

**Comment by champagnealexandre** [comment]
Seems to work for me, at least for Slack and Facebook Messenger. ✅

macOS: Sequoia 15.1.1 (just updated from 15.1)
alt-tab: 7.4.0 (just updated from 7.1.1)
Slack: 4.41.98
Messenger: 223.1.0.2.215

**Comment by SarcevicAntonio** [comment]
It always seems to work for me, but after ~30 minutes it breaks again 🙃 

macOS: 15.1.1
alt-tab: 7.4.0
Slack: 4.41.97


https://github.com/user-attachments/assets/3bc3f327-437e-48ec-8c24-bbd8f625f480

**Comment by filipef101** [comment]
I'm almost sure I replicated with Notes app, having two windows, it would never bring the main window to front. But accidentally force quitted the alt tab.. (apparently restarting alt tab even slack starts working again).

Could be worth anyone else when having this issue, trying on notes to confirm if it's  or not exclusive to electron apps

**Comment by champagnealexandre** [comment]
> It always seems to work for me, but after ~30 minutes it breaks again 🙃

Can confirm. Still an issue for me as well even though it briefly worked correctly.

**Comment by jonobri** [comment]
Can also confirm. Rebooting AltTab solves the problem, but only temporarily.

**Comment by porwoliktomas** [comment]
It's the same behavior for me. Restarting AltTab fixes the problem, but after a while, it stops working. Affected apps are Slack, Messenger, Notes, ChatGPT, Linear.

**Comment by DominikIlski** [comment]
Can confirm as well, the affected apps:
1. slack
1. messenger
1. telegram 

Issues started after the mac os update 
sequoia 15.2 
restarting the app solves the issue

**Comment by lwouis** [comment]
#4192
